kompari-tasks/src/optimizations.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2025 the Kompari Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use kompari::{list_image_dir, optimize_png, SizeOptimizationLevel};
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use std::cmp::min;
+use std::path::{Path, PathBuf};
+
+#[derive(Debug)]
+pub struct OptimizationResult {
+    pub path: PathBuf,
+    pub old_size: usize,
+    pub new_size: usize,
+    pub improvement: f32,
+    pub size_limit_breached: bool,
+    pub improvement_limit_breached: bool,
+}
+
+pub fn check_file_optimizations(
+    path: PathBuf,
+    args: &SizeCheckArgs,
+) -> kompari::Result<Option<OptimizationResult>> {
+    let old_data = std::fs::read(&path)?;
+    let old_size = old_data.len();
+    let new_data = optimize_png(old_data, SizeOptimizationLevel::High);
+    let new_size = min(new_data.len(), old_size);
+
+    let is_big = args.size_limit.is_some_and(|limit| {
+        if args.optimize {
+            new_size > limit * 1024 // Limit is given in KiB
+        } else {
+            old_size > limit * 1024
+        }
+    });
+
+    Ok(if old_size > new_size || is_big {
+        if args.optimize {
+            std::fs::write(&path, new_data)?;
+        }
+        let improvement = (old_size as f32 - new_size as f32) / old_size as f32;
+        Some(OptimizationResult {
+            path,
+            old_size,
+            new_size,
+            improvement,
+            size_limit_breached: is_big,
+            improvement_limit_breached: args
+                .improvement_limit
+                .is_some_and(|limit| limit < improvement),
+        })
+    } else {
+        None
+    })
+}
+
+pub fn check_size_optimizations(dir_path: &Path, args: &SizeCheckArgs) -> kompari::Result<()> {
+    let paths: Vec<_> = list_image_dir(dir_path)?.collect();
+    let progressbar = indicatif::ProgressBar::new(paths.len() as u64);
+    let results = paths
+        .into_par_iter()
```

Comment:
All the instances of `oxipng` which get started use the same thread pool, right? (i.e. they're not competing for cores)

Reply:
Yes, it should share the thread pool and generate same tasks for work stealing. At least as the Rayon version matches.